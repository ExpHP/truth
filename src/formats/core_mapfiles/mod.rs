use ::std::collections::BTreeMap;

use crate::game::Game;
use crate::eclmap::Eclmap;

pub mod anm;
pub mod msg;
pub mod std;

/// Struct for representing some embedded mapfile information (notably the default signatures and
/// register types for all mainstream games).
///
/// Basically, the current mapfile syntax has a bit of a problem that it is impossible to factor
/// out into reusable parts.  And since some formats frequently change signatures of individual
/// instructions, putting signatures there requires maintaining an unduly amount of copy-pasta.
///
/// This internal data structure is designed to be easier to refactor and maintain compared to
/// an actual mapfile, and in the future may even help guide attempts to solve the factoring
/// problem for user mapfile syntax.
pub struct CoreSignatures {
    /// A simple method of factoring out a single [`CoreSignatures`] into reusable
    /// portions when other methods fail to suffice. There's no specific, prescribed usage.
    ///
    /// Each item listed here will be applied in sequence before the definitions contained on
    /// the struct itself. (with later values overriding earlier ones)
    ///
    /// Note that a significant amount of boilerplate can be avoided by making use of the
    /// "minimum game" fields on [`Self::ins`] and [`Self::var`], and so some formats don't
    /// actually need to make use of this at all.
    inherit: &'static [&'static CoreSignatures],

    /// A set of tuples, each the addition of a new instruction or the change of an existing
    /// instruction's signature in some game.  (a `None` indicates removal; i.e. the jumptable
    /// entry now points to the default branch)
    ///
    /// The strings are mapfile string signatures.
    ///
    /// When converted to a mapfile, the program will run down this list and apply each item
    /// from the current game or earlier.  The presence of these "minimum game" fields allows
    /// a single [`CoreSignatures`] to be easily applied to an entire range of games.
    ins: &'static [(Game, u16, Option<&'static str>)],

    /// Like [`Self::ins`] but for registers.
    var: &'static [(Game, i32, Option<&'static str>)],
}

impl CoreSignatures {
    pub fn to_mapfile(&self, game: Game) -> Eclmap {
        let mut mapfile = Eclmap::new();
        self.apply_to_mapfile(game, &mut mapfile);
        mapfile
    }

    fn apply_to_mapfile(&self, game: Game, mapfile: &mut Eclmap) {
        for parent in self.inherit {
            parent.apply_to_mapfile(game, mapfile);
        }

        for &(min_game, opcode, siggy_str) in self.ins {
            if min_game <= game {
                insert_or_remove(&mut mapfile.ins_signatures, opcode as _, siggy_str);
            }
        }

        for &(min_game, reg_id, type_str) in self.var {
            if min_game <= game {
                insert_or_remove(&mut mapfile.gvar_types, reg_id as _, type_str);
            }
        }
    }
}

fn insert_or_remove<K, V>(map: &mut BTreeMap<K, V::Owned>, key: K, value: Option<&V>)
where
    K: Ord + Eq,
    V: ToOwned + ?Sized,
{
    if let Some(value) = value {
        map.insert(key, value.to_owned());
    } else {
        map.remove(&key);
    }
}
