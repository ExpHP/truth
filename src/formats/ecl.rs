use std::fmt;
use std::io;
use std::num::NonZeroU64;

use enum_map::EnumMap;
use indexmap::{IndexSet, IndexMap};

use crate::ast;
use crate::io::{BinRead, BinWrite, BinReader, BinWriter, Encoded, ReadResult, WriteResult, DEFAULT_ENCODING};
use crate::diagnostic::{Emitter};
use crate::error::{GatherErrorIteratorExt, ErrorReported};
use crate::game::Game;
use crate::ident::{Ident, ResIdent};
use crate::llir::{self, ReadInstr, RawInstr, InstrFormat, IntrinsicInstrKind};
use crate::meta::{self, FromMeta, FromMetaError, Meta, ToMeta};
use crate::pos::{Sp, Span};
use crate::value::{ScalarType};
use crate::context::CompilerContext;
use crate::passes::DecompileKind;
use crate::resolve::RegId;

// =============================================================================

/// Game-independent representation of an ANM file.
#[derive(Debug, Clone)]
pub struct OldeEclFile {
    pub subs: IndexMap<Ident, Vec<RawInstr>>,
    pub timelines: IndexMap<Ident, Vec<RawInstr>>,
    /// Filename of a read binary file, for display purposes only.
    binary_filename: Option<String>,
}

impl OldeEclFile {
    // pub fn decompile_to_ast(&self, game: Game, ctx: &mut CompilerContext, decompile_kind: DecompileKind) -> Result<ast::ScriptFile, ErrorReported> {
    //     let emitter = ctx.emitter.while_decompiling(self.binary_filename.as_deref());
    //     decompile(self, &emitter, &game_format(game), ctx, decompile_kind)
    // }
    //
    // pub fn compile_from_ast(game: Game, ast: &ast::ScriptFile, ctx: &mut CompilerContext) -> Result<Self, ErrorReported> {
    //     compile(&game_format(game), ast, ctx)
    // }

    pub fn write_to_stream(&self, w: &mut BinWriter, game: Game) -> WriteResult {
        let emitter = w.emitter();
        write_olde_ecl(w, &emitter, &game_format(game)?, self)
    }

    pub fn read_from_stream(r: &mut BinReader, game: Game) -> ReadResult<Self> {
        let emitter = r.emitter();
        read_olde_ecl(r, &emitter, &game_format(game)?)
    }
}

// =============================================================================

#[cfg(nope)]
fn decompile(
    anm_file: &OldeEclFile,
    emitter: &impl Emitter,
    format: &OldeFileFormat,
    ctx: &mut CompilerContext,
    decompile_kind: DecompileKind,
) -> Result<ast::ScriptFile, ErrorReported> {

}

// =============================================================================

#[cfg(nope)]
fn compile(
    format: &OldeFileFormat,
    ast: &ast::ScriptFile,
    ctx: &mut CompilerContext,
) -> Result<OldeEclFile, ErrorReported> {

}

// =============================================================================

fn read_olde_ecl(
    reader: &mut BinReader,
    emitter: &impl Emitter,
    format: &OldeFileFormat,
) -> ReadResult<OldeEclFile> {
    let instr_format = &*format.instr_format();
    let timeline_format = &*format.timeline_format();

    let start_pos = reader.pos()?;

    if let Some(magic) = format.magic() {
        reader.expect_magic(emitter, &magic.to_le_bytes())?;
    }

    let num_subs = reader.read_u16()? as usize;
    let num_subs_high_word = reader.read_u16()? as usize;

    let timeline_array_len; // how many u32s to read for timelines
    let expected_nonzero_timelines; // how many of them SHOULD be nonzero
    match format.timeline_array_kind() {
        TimelineArrayKind::Pofv => {
            timeline_array_len = num_subs_high_word;
            expected_nonzero_timelines = num_subs_high_word;
        },
        TimelineArrayKind::Pcb { cap } => {
            timeline_array_len = cap;
            expected_nonzero_timelines = num_subs_high_word + 1;
        }
        TimelineArrayKind::Eosd { cap } => {
            timeline_array_len = cap;
            expected_nonzero_timelines = 1;
            if num_subs_high_word != 0 {
                emitter.emit(warning!(
                    "unexpected nonzero high word for num_subs: {:#x}", num_subs_high_word,
                )).ignore();
            }
        },
    };
    let timeline_offsets = reader.read_u32s(timeline_array_len)?;
    let sub_offsets = reader.read_u32s(num_subs)?;

    // expect a prefix of the timeline array to be filled
    // (because we can't represent a sparsely filled timeline array in the AST)
    let mut num_timelines = timeline_offsets.iter().position(|&x| x == 0).unwrap_or(timeline_offsets.len());
    for &offset in &timeline_offsets[num_timelines..] {
        if offset != 0 {
            return Err(emitter.emit(error!("unexpected timeline offset {:#x} after a null offset", offset)));
        }
    }

    if num_timelines != expected_nonzero_timelines {
        emitter.emit(warning!(
            "expected {} nonzero entries in timeline table, but found {}",
            expected_nonzero_timelines, num_timelines,
        )).ignore();
    }
    if matches!(format.timeline_array_kind(), TimelineArrayKind::Pcb { .. }) {
        num_timelines -= 1;  // in these games, that last entry points to the end of the file
    }

    eprintln!("NUM_SUBS {}", num_subs);
    eprintln!("NUM_TIMELINES {}", num_timelines);

    let subs = sub_offsets.into_iter().enumerate().map(|(index, sub_offset)| {
        let name = auto_sub_name(index as u32);

        eprintln!(": SUB {} AT {:#x}", index, sub_offset);
        reader.seek_to(start_pos + sub_offset as u64)?;
        let instrs = emitter.chain_with(|f| write!(f, "in sub {}", index), |emitter| {
            eprintln!("    TIME _OP_ SIZE DIFF MASK  ARGS");
            llir::read_instrs(reader, emitter, instr_format, sub_offset as u64, None)
        })?;
        Ok((name, instrs))
    }).collect::<ReadResult<_>>()?;

    let timelines = timeline_offsets[..num_timelines].iter().enumerate().map(|(index, &sub_offset)| {
        let name = auto_timeline_name(index as u32);

        eprintln!(": TIMELINE {} AT {:#x}", index, sub_offset);
        reader.seek_to(start_pos + sub_offset as u64)?;
        let instrs = emitter.chain_with(|f| write!(f, "in timeline sub {}", index), |emitter| {
            eprintln!("TIME ARG0 _OP_ SZ DF  ARGS");
            llir::read_instrs(reader, emitter, timeline_format, sub_offset as u64, None)
        })?;
        Ok((name, instrs))
    }).collect::<ReadResult<_>>()?;

    let binary_filename = Some(reader.display_filename().to_string());
    Ok(OldeEclFile { subs, timelines, binary_filename })
}

pub fn auto_sub_name(i: u32) -> Ident {
    format!("sub{}", i).parse::<Ident>().unwrap()
}
pub fn auto_timeline_name(i: u32) -> Ident {
    format!("timeline{}", i).parse::<Ident>().unwrap()
}

fn write_olde_ecl(
    w: &mut BinWriter,
    emitter: &impl Emitter,
    format: &OldeFileFormat,
    ecl: &OldeEclFile,
) -> WriteResult {
    let instr_format = &*format.instr_format();
    let timeline_format = &*format.timeline_format();

    let start_pos = w.pos()?;

    if let Some(magic) = format.magic() {
        w.write_u32(magic)?;
    }

    if ecl.timelines.len() > format.timeline_array_kind().max_timelines() {
        // FIXME: NEEDSTEST for each game
        return Err(emitter.emit(error!("too many timelines! (max allowed in this game is {})", ecl.timelines.len())));
    }

    match format.timeline_array_kind() {
        | TimelineArrayKind::Pofv { .. }
        | TimelineArrayKind::Pcb { .. } => {
            w.write_u16(ecl.subs.len() as _)?;
            w.write_u16(ecl.timelines.len() as _)?;
        },
        | TimelineArrayKind::Eosd { .. } => {
            w.write_u16(ecl.subs.len() as _)?;
            w.write_u16(0)?;
        },
    };

    let timeline_array_len = match format.timeline_array_kind() {
        TimelineArrayKind::Pcb { cap } => cap,
        TimelineArrayKind::Eosd { cap, .. } => cap,
        TimelineArrayKind::Pofv => ecl.timelines.len(),
    };

    let script_offsets_pos = w.pos()?;
    for _ in 0..ecl.subs.len() + timeline_array_len {
        w.write_u32(0)?;
    }

    let mut sub_offsets = vec![];
    for (index, (ident, sub)) in ecl.subs.iter().enumerate() {
        sub_offsets.push(w.pos()? - start_pos);
        emitter.chain_with(|f| write!(f, "in sub {} (index {})", ident, index), |emitter| {
            llir::write_instrs(w, emitter, instr_format, &sub)
        })?;
    }

    let mut timeline_offsets = vec![];
    for (index, (ident, sub)) in ecl.timelines.iter().enumerate() {
        timeline_offsets.push(w.pos()? - start_pos);
        emitter.chain_with(|f| write!(f, "in sub {} (index {})", ident, index), |emitter| {
            llir::write_instrs(w, emitter, timeline_format, &sub)
        })?;
    }

    let end_pos = w.pos()?;
    if matches!(format.timeline_array_kind(), TimelineArrayKind::Pcb { .. }) {
        // in these games, that last entry points to the end of the file
        timeline_offsets.push(end_pos - start_pos);
    }
    timeline_offsets.resize(timeline_array_len, 0);

    w.seek_to(script_offsets_pos)?;
    for offset in timeline_offsets {
        w.write_u32(offset as u32)?;
    }
    for offset in sub_offsets {
        w.write_u32(offset as u32)?;
    }
    w.seek_to(end_pos)?;
    Ok(())
}

// =============================================================================

fn game_format(game: Game) -> Result<OldeFileFormat, ErrorReported> {
    match game {
        | Game::Th06 | Game::Th07 | Game::Th08 | Game::Th09 | Game::Th095
        => Ok(OldeFileFormat { game }),

        _ => Err(unimplemented!("game {} not yet supported", game)),
    }
}

pub fn game_core_mapfile(game: Game) -> crate::Eclmap {
    super::core_mapfiles::msg::core_signatures(game).to_mapfile(game)
}

// =============================================================================

struct OldeFileFormat { game: Game }

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
enum TimelineArrayKind {
    /// There appears to be space for `cap` items, but only the first is used.
    Eosd { cap: usize },
    /// The timeline offset array is a fixed size, but the file specifies how many are actually used.
    ///
    /// (there will be that many nonzero entries *plus one*; the final one will be the ending offset of
    ///  the last timeline, which should coincide with the size of the file)
    Pcb { cap: usize },
    /// The length of the timeline offset array is specified in the file.  All entries are used.
    Pofv,
}

impl TimelineArrayKind {
    fn max_timelines(&self) -> usize {
        match *self {
            TimelineArrayKind::Eosd { .. } => 1,
            TimelineArrayKind::Pcb { cap } => cap - 1, // room for file size
            TimelineArrayKind::Pofv => usize::MAX,
        }
    }
}

impl OldeFileFormat {
    fn magic(&self) -> Option<u32> {
        match self.game {
            | Game::Th06 | Game::Th07 => None,
            | Game::Th08 | Game::Th095 => Some(0x00_00_08_00),
            | Game::Th09 => Some(0x00_00_09_00),
            _ => unimplemented!("game not yet supported"),
        }
    }

    fn timeline_array_kind(&self) -> TimelineArrayKind {
        match self.game {
            | Game::Th06 => TimelineArrayKind::Eosd { cap: 3 },

            | Game::Th07 | Game::Th08 | Game::Th095
            => TimelineArrayKind::Pcb { cap: 16 },

            | Game::Th09 => TimelineArrayKind::Pofv,

            _ => unimplemented!("game not yet supported"),
        }
    }

    fn instr_format(&self) -> Box<dyn InstrFormat> { Box::new(InstrFormat06) }
    fn timeline_format(&self) -> Box<dyn InstrFormat> { Box::new(TimelineInstrFormat { game: self.game }) }
}

struct InstrFormat06;

impl InstrFormat for InstrFormat06 {
    fn intrinsic_opcode_pairs(&self) -> Vec<(llir::IntrinsicInstrKind, u16)> {
        vec![]
    }

    fn instr_header_size(&self) -> usize { 12 }

    fn read_instr(&self, f: &mut BinReader, _: &dyn Emitter) -> ReadResult<ReadInstr> {
        let time = f.read_i32()?;
        let opcode = f.read_u16()?;
        let size = f.read_u16()? as usize;
        let difficulty = f.read_u16()?;
        let param_mask = f.read_u16()?;

        let args_blob = f.read_byte_vec(size - self.instr_header_size())?;

        eprintln!("{:08x} {:04x} {:04x} {:04x} {:04x}  {}", time, opcode, size, difficulty, param_mask, crate::io::hexify(&args_blob));
        let instr = RawInstr { time, opcode, param_mask, args_blob, difficulty, ..RawInstr::DEFAULTS };

        if opcode == (-1_i16) as u16 {
            Ok(ReadInstr::Terminal)
        } else {
            Ok(ReadInstr::Instr(instr))
        }
    }

    fn write_instr(&self, f: &mut BinWriter, _: &dyn Emitter, instr: &RawInstr) -> WriteResult {
        f.write_i32(instr.time)?;
        f.write_u16(instr.opcode)?;
        f.write_u16(self.instr_size(instr) as _)?;
        f.write_u16(instr.difficulty)?;
        f.write_u16(instr.param_mask)?;
        f.write_all(&instr.args_blob)?;
        Ok(())
    }

    fn write_terminal_instr(&self, f: &mut BinWriter, _: &dyn Emitter) -> WriteResult {
        f.write_i32(-1)?; // time
        f.write_i16(-1)?; // opcode
        f.write_u16(self.instr_header_size() as _)?; // size
        f.write_u16(0xff00)?; // difficulty
        f.write_u16(0x00ff)?; // param_mask
        Ok(())
    }
}

struct TimelineInstrFormat { game: Game }
impl TimelineInstrFormat {
    /// In some games, the terminal instruction is shorter than an actual instruction.
    fn has_short_terminal(&self) -> bool {
        self.game < Game::Th08
    }
}

impl InstrFormat for TimelineInstrFormat {
    fn intrinsic_opcode_pairs(&self) -> Vec<(llir::IntrinsicInstrKind, u16)> {
        vec![]
    }

    fn instr_header_size(&self) -> usize { 8 }

    fn read_instr(&self, f: &mut BinReader, emitter: &dyn Emitter) -> ReadResult<ReadInstr> {
        let time = f.read_i16()? as i32;
        let arg_0 = f.read_i16()? as i32;

        // with some games the terminal instruction is only 4 bytes long so we must check now
        if self.has_short_terminal() && (time, arg_0) == (-1, 4) {
            // FIXME: really should be MaybeTerminal since both arg_0 = 4 and time = -1
            //        are things that naturally occur on real instructions.
            return Ok(ReadInstr::Terminal);
        }

        let opcode = f.read_u16()?;
        let size = f.read_u8()? as usize;
        let difficulty = f.read_u8()? as u16;

        // in games with the normal-sized terminal, the size is incorrectly 0, so check before reading args
        if !self.has_short_terminal() && (time, arg_0, opcode, size, difficulty) == (-1, -1, 0, 0, 0) {
            // FIXME: really should be MaybeTerminal
            return Ok(ReadInstr::Terminal);
        }

        let args_blob = f.read_byte_vec(size - self.instr_header_size())?;
        eprintln!("{:04x} {:04x} {:04x} {:02x} {:02x}  {}", time as i16, arg_0 as u16, opcode as u16, size as u8, difficulty as u8, crate::io::hexify(&args_blob));

        let instr = RawInstr {
            time, opcode, difficulty, args_blob,
            extra_arg: Some(arg_0 as _),
            ..RawInstr::DEFAULTS
        };
        Ok(ReadInstr::Instr(instr))
    }

    fn write_instr(&self, f: &mut BinWriter, _: &dyn Emitter, instr: &RawInstr) -> WriteResult {
        f.write_i16(instr.time as _)?;
        f.write_i16(instr.extra_arg.expect("missing extra_arg on timeline?!") as _)?;
        f.write_u16(instr.opcode)?;
        f.write_u8(self.instr_size(instr) as _)?;
        f.write_u8(instr.difficulty as _)?;
        f.write_all(&instr.args_blob)?;
        Ok(())
    }

    fn write_terminal_instr(&self, f: &mut BinWriter, _: &dyn Emitter) -> WriteResult {
        if self.game < Game::Th08 {
            f.write_i16(-1)?; // time
            f.write_i16(4)?; // arg_0
        } else {
            f.write_i16(-1)?; // time
            f.write_i16(-1)?; // arg_0
            f.write_u32(0)?; // opcode, size, difficulty
        }
        Ok(())
    }
}
