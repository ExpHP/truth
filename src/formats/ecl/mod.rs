use crate::ast;
use crate::io::{BinReader, BinWriter, ReadResult, WriteResult};
use crate::error::{ErrorReported};
use crate::game::{Game};
use crate::llir::{DecompileOptions};
use crate::context::CompilerContext;

use ecl_06::OldeEclFile;
pub mod ecl_06;

use ecl_10::StackEclFile;
pub mod ecl_10;

#[derive(Debug, Clone)]
#[derive(strum::EnumDiscriminants)]
#[strum_discriminants(name(EclFileKind))]
pub enum EclFile {
    Olde(ecl_06::OldeEclFile),
    Stack(ecl_10::StackEclFile),
}

impl EclFile {
    pub fn decompile_to_ast(&self, game: Game, ctx: &mut CompilerContext, decompile_options: &DecompileOptions) -> Result<ast::ScriptFile, ErrorReported> {
        match self {
            EclFile::Olde(ecl) => ecl.decompile_to_ast(game, ctx, decompile_options),
            EclFile::Stack(ecl) => ecl.decompile_to_ast(game, ctx, decompile_options),
        }
    }

    pub fn compile_from_ast(game: Game, ast: &ast::ScriptFile, ctx: &mut CompilerContext) -> Result<Self, ErrorReported> {
        match EclFileKind::of_game(game) {
            EclFileKind::Olde => OldeEclFile::compile_from_ast(game, ast, ctx).map(EclFile::Olde),
            EclFileKind::Stack => StackEclFile::compile_from_ast(game, ast, ctx).map(EclFile::Stack),
        }
    }

    pub fn write_to_stream(&self, w: &mut BinWriter, game: Game) -> WriteResult {
        match self {
            EclFile::Olde(ecl) => ecl.write_to_stream(w, game),
            EclFile::Stack(ecl) => ecl.write_to_stream(w, game),
        }
    }

    pub fn read_from_stream(r: &mut BinReader, game: Game) -> ReadResult<Self> {
        match EclFileKind::of_game(game) {
            EclFileKind::Olde => OldeEclFile::read_from_stream(r, game).map(EclFile::Olde),
            EclFileKind::Stack => StackEclFile::read_from_stream(r, game).map(EclFile::Stack),
        }
    }
}

impl EclFileKind {
    pub fn of_game(game: Game) -> Self {
        if game < Game::Th10 {
            EclFileKind::Olde
        } else {
            EclFileKind::Stack
        }
    }
}
