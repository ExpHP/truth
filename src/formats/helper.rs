use std::collections::HashSet;

use crate::ast;
use crate::pos::Sp;
use crate::ident::Ident;
use crate::diagnostic::Emitter;
use crate::error::{ErrorReported, ErrorFlag};

/// A structure representing the union of the text file layouts for all scripting languages.
///
/// After script is parsed into a `Vec<`[`ast::Item`]`>`, it is quickly converted into one of these.
/// This process consolidates a lot of code that e.g. looks for duplicate `script` names or errors
/// on unsupported syntax in a file.  By fully destructuring it in code, you can get "unused" warnings
/// which will help reduce bugs where a format silently ignores a certain type of item.
///
/// When [printed][`crate::fmt`] or [traversed][`crate::ast::Visit`], the items may print in a different
/// order from the original AST.
#[derive(Debug, Clone, Default)]
pub struct ScriptFile {
    pub meta: Option<ast::ItemMeta>,
    pub toplevel_consts: Vec<ast::ItemConstVar>,
    /// `script`s not underneath an `entry`.
    pub toplevel_scripts: Vec<ast::ItemScript>,
    /// Functions not defined inside a block of code.
    pub toplevel_funcs: Vec<ast::ItemFunc>,
    pub mapfiles: Vec<ast::ItemPragma>,
    pub image_sources: Vec<ast::ItemPragma>,
    pub entries: Vec<Entry>,
}

#[derive(Debug, Clone)]
pub struct Entry {
    pub meta: ast::ItemMeta,
    pub scripts: Vec<ast::ItemScript>,
}

impl ScriptFile {
    pub fn from_ast(items: Vec<Sp<ast::Item>>, emitter: &impl Emitter) -> Result<ScriptFile, ErrorReported> {
        let mut errors = ErrorFlag::new();
        let mut output = ScriptFile::default();
        let mut current_entry = None;

        let mut script_checker = RedefinitionChecker::new(emitter, "script");
        let mut function_checker = RedefinitionChecker::new(emitter, "function");

        for item in items {
            match item.value {
                ast::Item::Pragma(pragma@ast::ItemPragma { kind: sp_pat!(ast::PragmaKindKeyword::Mapfile), .. }) => {
                    output.mapfiles.push(pragma);
                },
                ast::Item::Pragma(pragma@ast::ItemPragma { kind: sp_pat!(ast::PragmaKindKeyword::ImageSource), .. }) => {
                    output.image_sources.push(pragma);
                },
                ast::Item::Meta(meta@ast::ItemMeta { keyword: sp_pat![kw_span => token![meta]], .. }) => {
                    if let Some(prev_meta) = output.meta.replace(meta) {
                        errors.set(emitter.emit(error!(
                            message("'meta' supplied multiple times"),
                            secondary(prev_meta.keyword, "previously supplied here"),
                            primary(kw_span, "duplicate 'meta'"),
                        )));
                    }
                },
                ast::Item::Meta(meta@ast::ItemMeta { keyword: sp_pat![kw_span => token![entry]], .. }) => {
                    // Begin a new entry
                    if let Some(prev_entry) = current_entry.take() {
                        output.entries.push(prev_entry);
                    }
                    current_entry = Some(Entry { meta, scripts: vec![] });
                },
                ast::Item::Script(script) => {
                    if let Err(e) = script_checker.check_for_redefinition(&script.ident) {
                        errors.set(e);
                    }
                    match current_entry.as_mut() {
                        Some(entry) => entry.scripts.push(script),
                        None => output.toplevel_scripts.push(script),
                    }
                },
                ast::Item::ConstVar(item) => {
                    output.toplevel_consts.push(item);
                },
                ast::Item::Func(item) => {
                    if let Err(e) = function_checker.check_for_redefinition(&sp!(item.ident.span => item.ident.as_raw().clone())) {
                        errors.set(e);
                    }
                    output.toplevel_funcs.push(item);
                },
            }
        }

        errors.into_result(output)
    }
}

struct RedefinitionChecker<'a> {
    emitter: &'a dyn Emitter,
    noun: &'a str,
    seen: HashSet<Sp<Ident>>,
}

impl<'a> RedefinitionChecker<'a> {
    fn new(emitter: &'a dyn Emitter, noun: &'a str) -> Self {
        RedefinitionChecker { emitter, noun, seen: Default::default() }
    }

    fn check_for_redefinition(&mut self, ident: &Sp<Ident>) -> Result<(), ErrorReported> {
        if let Some(prev) = self.seen.replace(ident.clone()) {
            return Err(self.emitter.as_sized().emit(error!(
                message("redefinition of {} '{ident}'", self.noun),
                primary(ident, ""),
                secondary(prev, "previously defined here"),
            )));
        }
        Ok(())
    }
}

pub fn meta_not_supported(emitter: &impl Emitter, meta: &Option<ast::ItemMeta>) -> Result<(), ErrorReported> {
    if let Some(ast::ItemMeta { keyword, .. }) = meta {
        return Err(emitter.emit(error!(
            message("unexpected '{keyword}'"),
            primary(keyword, "not valid in this format"),
        )))
    }
    Ok(())
}

pub fn scripts_not_supported(emitter: &impl Emitter, scripts: &[ast::ItemScript], note: Option<String>) -> Result<(), ErrorReported> {
    if let Some(ast::ItemScript { keyword, .. }) = scripts.iter().next() {
        let mut diag = error!(
            message("unexpected 'script'"),
            primary(keyword, "not valid in this format"),
        );
        if let Some(note) = note {
            diag.note(note);
        }
        return Err(emitter.emit(diag));
    }
    Ok(())
}

pub fn functions_not_supported(emitter: &impl Emitter, funcs: &[ast::ItemFunc]) -> Result<(), ErrorReported> {
    if let Some(ast::ItemFunc { ty_keyword, .. }) = funcs.iter().next() {
        return Err(emitter.emit(error!(
            message("unexpected function definition"),
            primary(ty_keyword, "not valid in this format"),
        )))
    }
    Ok(())
}

pub fn pragma_not_supported(emitter: &impl Emitter, pragmas: &[ast::ItemPragma], note: Option<String>) -> Result<(), ErrorReported> {
    if let Some(ast::ItemPragma { kind, .. }) = pragmas.iter().next() {
        return Err(emitter.emit(error!(
            message("unexpected 'pragma {kind}'"),
            primary(kind, "not valid in this format"),
        )));
    }
    Ok(())
}
