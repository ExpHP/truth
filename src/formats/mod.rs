pub mod anm;
pub mod std;
pub mod msg;
pub mod mission;
pub mod ecl;

mod core_mapfiles;
