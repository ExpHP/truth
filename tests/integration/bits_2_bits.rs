//! A test that decompiles and recompile a file to test that the result is bit-for-bit identical.
//!
//! This is primarily for regression testing against changes that would break bit-for-bit recompilation.
//!
//! Typically used via the [`b2b_test`] macro which also does an insta snapshot of the text; this
//! can give clues on possible causes of trouble.
//!
//! Tests for things like loop- and if- decompilation (which used to be here) are better suited to the
//! [`Self::sbsb_test`], which start from a readable script file.

use crate::integration_impl::formats::*;

#[track_caller]
fn file(fname: &str) -> std::path::PathBuf {
    let mut path = std::path::PathBuf::from("tests/integration/bits-2-bits");
    path.push(fname);
    assert!(path.exists());
    path
}

macro_rules! b2b_test {
    (   $format:expr, $map:expr, $test_name:ident, $fname:literal
        // Expected strings, all of which must be contained somewhere in the output for the test to succeed.
        // These can serve to help clarify what the test is actually testing for, and they serve as an
        // additional speed-bump against accidentally accepting a broken snapshot.
        //
        // Some tests don't have any expected strings; those tests are most likely only concerned that the
        // data can be round-tripped back to binary.
        $(, expected=$expected:literal)*
        $(,)?
    ) => {
        #[test]
        fn $test_name() {
            $format.bits_to_bits(file($fname), $map, |s| {
                $( assert!(s.contains($expected)); )*
                insta::assert_snapshot!(s)
            });
        }
    }
}

// a dumb test to make sure that we don't just dump files in the input directory and
// forget to add tests for them.
#[test]
fn all_files_tested() {
    let mut not_mentioned = vec![];
    let this_file_src = std::fs::read_to_string(file!()).unwrap();
    for entry in std::fs::read_dir("tests/integration/bits-2-bits").unwrap() {
        let path = entry.unwrap().path();
        if path.is_file() {
            let file_name = path.file_name().unwrap().to_str().unwrap();

            if !this_file_src.contains(file_name) {
                not_mentioned.push(file_name.to_string());
            }
        }
    }

    if !not_mentioned.is_empty() {
        panic!("Files not mentioned in {}: {:?}", file!(), not_mentioned);
    }
}

// STD metadata.
b2b_test!(STD_08, "map/any.stdm", std08_rects, "th08-rects.std");
b2b_test!(STD_08, "map/any.stdm", std08_strip, "th08-strip.std");
b2b_test!(STD_08, "map/any.stdm", std08_instance_order, "th08-instance-order.std");
b2b_test!(STD_08, "map/any.stdm", std08_nonzero_padding, "th08-nonzero-padding.std");
b2b_test!(STD_08, "map/any.stdm", std08_empty_script, "th08-empty-script.std");
b2b_test!(STD_06, "map/any.stdm", std06_general, "th06-general.std");
b2b_test!(STD_12, "map/any.stdm", std12_general, "th12-general.std");

// Test that named regs use are named in the output.
b2b_test!(
    ANM_12, "map/any.anmm", anm12_registers, "th12-registers.anm",
    expected="I1",
);
// Decompilation of instructions with no signature.
b2b_test!(
    ANM_12, file("th12-unknown-signature.anmm"), anm12_unknown_signature, "th12-unknown-signature.anm",
    expected="hasNameButNoSignature",  // there's a named instr with no signature in the mapfile...
    expected="ins_1011",  // ...as well as an unnamed one.
);

// Decompilation of word-sized arguments.
b2b_test!(ANM_12, "map/any.anmm", anm12_anchor_ss_signature, "th12-anchor-ss-signature.anm");

// Decompilation of sprite/script args to use names when appropriate.
b2b_test!(
    ANM_12, "map/any.anmm", anm12_sprite_script_args, "th12-sprite-script-args.anm",
    expected="scriptNew(script1)",
    expected="sprite(sprite0)",
);
b2b_test!(
    ANM_12, "map/any.anmm", anm12_sprite_unset, "th12-sprite-unset.anm",
    expected="sprite(-1)",
);
b2b_test!(
    ANM_12, "map/any.anmm", anm12_sprite_non_sequential, "th12-sprite-non-sequential.anm",
    expected="sprite47",
);
b2b_test!(
    ANM_12, "map/any.anmm", anm12_sprite_duplicates, "th12-sprite-duplicates.anm",
);

// MSG script table funny business.
b2b_test!(MSG_06, "map/any.msgm", msg06_multiple_scripts, "th06-multiple-scripts.msg");

// Decoding of string arguments.
b2b_test!(MSG_06, "map/any.msgm", msg06_z_string, "th06-z-string.msg");
b2b_test!(MSG_08, "map/any.msgm", msg08_m_string, "th08-m-string.msg");
b2b_test!(MSG_09, "map/any.msgm", msg09_m_string, "th09-m-string.msg");
b2b_test!(MSG_09, "map/any.msgm", msg09_default_mismatched_flags, "th09-default-mismatched-flags.msg");
b2b_test!(MSG_09, "map/any.msgm", msg09_default_zero_with_flags, "th09-default-zero-with-flags.msg");
b2b_test!(MSG_09, "map/any.msgm", msg09_default_zero, "th09-default-zero.msg");
b2b_test!(MSG_09, "map/any.msgm", msg09_no_default, "th09-no-default.msg");

// TH11 MSG has furigana but does not yet have the furigana encoding bug in masked strings.
b2b_test!(MSG_11, "map/any.msgm", msg11_furibug_not_applicable, "th11-furibug-not-applicable.msg");
// Typical furigana bug example.
b2b_test!(MSG_12, "map/any.msgm", msg12_furibug, "th12-furibug.msg");
// Example with particular string length that caused a bug in TH17 Extra.
b2b_test!(MSG_17, "map/any.msgm", msg17_furibug_ex_regression, "th17-furibug-ex-regression.msg");
