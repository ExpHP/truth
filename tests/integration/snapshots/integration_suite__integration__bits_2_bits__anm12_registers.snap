---
source: tests/integration/bits_2_bits.rs
expression: s
---
#pragma mapfile "map/any.anmm"

entry {
    path: "subdir/file.png",
    source: "none",
    buf_width: 512,
    buf_height: 512,
    buf_format: 3,
    offset_x: 0,
    offset_y: 0,
    colorkey: 0x0,
    memory_priority: 0,
    low_res_scale: false,
    sprites: {sprite0: {x: 0.0, y: 0.0, w: 512.0, h: 480.0}},
}


script 0 script0 {
    I0 = 3;
    I1 = I0;
}

